### DISCUSSION TITLE
JS memory error when using BriefcaseDb.Elements.getAspects method

### DISCUSSION THREAD

**Discussion by nn15296** [iTwin/itwinjs-core#236] [Q&A]
When using the getAspects() method from the BriefcaseDb.Elements class (link to documentation: https://www.itwinjs.org/reference/imodeljs-backend/imodels/briefcasedb.elements/getaspects/), it will throw a JS heap out of memory error on certain elements (see picture below). Having looked at one of these elements, it isn't obvious to me how the element differs from the rest; it's not got fields with noticeably larger strings for example. Any ideas as to what's causing this and how to resolve it? Thanks.

![image](https://user-images.githubusercontent.com/33571096/99245913-5074b600-27fc-11eb-9dd6-b4bb60dd1fb6.png)

**Comment by pmconne** [comment]
Looks like base64 encoding of the element's geometry stream may be to blame ("Very long string[19622645]"). @scsewall?

**Reply by scsewall** [reply]
@nn15296 When complex geometry gets represented in string form, the string can get quite large. In most cases, however, it is not required to ever get serialized as a string. Is there something special about your use case?

Could you please give me some details on what your program is doing? For example, is it a backend-only batch program visiting every Element or is it an interactive web application? Also, what version of the iModel.js libraries are you using?

**Reply by nn15296** [reply]
@scsewall Thanks for taking a look. It's a backend program which visits each element and obtains the element aspect info. For most elements, this error doesn't occur. I've seen the geometry stream (I can definitely see why it would cause a memory error) for other elements and I don't need it but not sure how to exclude it when using getAspects(). I'm using versions 2.3.0.

**Reply by scsewall** [reply]
@nn15296 We need to figure out why your problem is running out of JavaScript memory. A few questions and comments:

- Are you doing anything that accumulates information per Element that you visit? 
- If you are visiting each source Element to import it into a new target iModel, are you calling `saveChanges` on a regular interval? I would recommend something like every 10,000 elements or so.
- Another possibility is there is a problem Element or ElementAspect. Are you logging Ids to determine where you are failing? If the failure is consistent then we will need to look into the particular Element or ElementAspect.
- Is there any reason why you are hanging back on version 2.3? It will be much easier for us to look into problems if you are using a more current version of the iModel.js libraries.

**Reply by nn15296** [reply]
@scsewall  There isn't any information accumulation as far as I'm aware. It seems to only happen on a specific element. If I try and call getAspects() on this element, it will crash even if the program doesn't iterate through the other elements. I know the particular element and elementAspect which causes it to crash. I can query this element/elementAspect using an inner join in imodelconsole, and it loads the information with no problems. The GeometryStream field is set as NULL so it's not clear which field is causing the problem. There's no reason for me to hang back on version 2.3, so I'm happy to update if that will help.

**Reply by nn15296** [reply]
@scsewall I've checked the elements and they are very detailed with curves and intricate features (one of them is a sink for example). Therefore it seems likely that the geometry stream will be much larger than that of other elements. As I mentioned in the previous comment, when querying using imodelconsole, it doesn't crash and the geometrystream field is set as NULL, which may be a way of mitigating this problem. Is there any way we can replicate this when using the getAspects() method because we don't need the geometry stream field so having it as NULL is not an issue.
